core/event.cpp

```diff
@@ -93,5 +96,138 @@ Variable broadcast_weights(const DataArrayConstView &events) {
+/// than the typical logarithmic growth. This yields a 5x speedup in some cases,
+/// without apparent negative effect on the other cases.
+void reserve(const VariableView &events, const VariableConstView &capacity) {
+  transform_in_place<
+      pair_custom_t<std::pair<event_list<double>, scipp::index>>,
+      pair_custom_t<std::pair<event_list<float>, scipp::index>>,
+      pair_custom_t<std::pair<event_list<int64_t>, scipp::index>>,
+      pair_custom_t<std::pair<event_list<int32_t>, scipp::index>>>(
+      events, capacity,
+      overloaded{[](auto &&sparse_, const scipp::index capacity_) {
+                   if (capacity_ > 2 * scipp::size(sparse_))
+                     return sparse_.reserve(capacity_);
+                 },
+                 transform_flags::expect_no_variance_arg<1>,
+                 [](const units::Unit &, const units::Unit &) {}});
+}
+
+namespace filter_detail {
+template <class T>
+using make_select_args = std::tuple<event_list<T>, span<const T>>;
+template <class T, class Index>
+using copy_if_args = std::tuple<event_list<T>, event_list<Index>>;
+
+/// Return new variable with values copied from `var` if index is included in
+/// `select`.
+constexpr auto copy_if = [](const VariableConstView &var,
+                            const VariableConstView &select) {
+  return transform<std::tuple<
+      copy_if_args<double, int32_t>, copy_if_args<float, int32_t>,
+      copy_if_args<int64_t, int32_t>, copy_if_args<int32_t, int32_t>,
+      copy_if_args<double, int64_t>, copy_if_args<float, int64_t>,
+      copy_if_args<int64_t, int64_t>, copy_if_args<int32_t, int64_t>>>(
+      var, select,
+      overloaded{
+          transform_flags::expect_no_variance_arg<1>,
+          [](const auto &var_, const auto &select_) {
+            using VarT = std::decay_t<decltype(var_)>;
+            using Events = event_list<typename VarT::value_type>;
+            const auto size = scipp::size(select_);
+            if constexpr (detail::is_ValuesAndVariances_v<VarT>) {
+              std::pair<Events, Events> out;
+              out.first.reserve(size);
+              out.second.reserve(size);
+              for (const auto i : select_) {
+                out.first.push_back(var_.values[i]);
+                out.second.push_back(var_.variances[i]);
+              }
+              return out;
+            } else {
+              Events out;
+              out.reserve(size);
+              for (const auto i : select_)
+                out.push_back(var_[i]);
+              return out;
+            }
+          },
+          [](const units::Unit &var_, const units::Unit &) { return var_; }});
+};
+
+/// Return list of indices with coord values for given dim inside interval.
+template <class T>
+const auto make_select = [](const DataArrayConstView &array, const Dim dim,
+                            const VariableConstView &interval) {
+  return transform<
+      std::tuple<make_select_args<double>, make_select_args<float>,
```

Comment:
So currently you can only filter on numeric types.
Any use case where one might want to filter on something else? I guess not really the way it's implemented, using `>=` and `<` comparisons.
Maybe you would then just make a mask if you wanted to filter on something else.
So filtering on dates would just be done using integers that convert the epochs?

Reply:
This is filtering only for event data, so filtering is only possible for things that we have event lists of. There is `filter` for `Variable` that may be able to handle the cases you are asking for.

Regarding filtering on things like `datetime64`, I suppose that is essentially also a "numeric" type so support would need to be added here.